### ISSUE TITLE
Web front-end shows blank page

### ISSUE THREAD

**Issue by aujawindar** [YunoHost-Apps/mastodon_ynh#341]
### Describe the bug
As [two different](https://forum.yunohost.org/t/mastodon-web-client-shows-blank-page/21500) [reports](https://forum.yunohost.org/t/mastodon-inaccessible-par-le-web/16025/) on forum.yunohost.org show, the Mastodon web client shows a blank page and cannot be used. [Example here](https://mastodon.fjerland.no/). The page source states that Javascript should be enabled, which it is.

### Context

- Hardware: Laptop 
- YunoHost version:  11.0.10.1 
- I have access to my server: Through SSH and through the webadmin
- Are you in a special context or did you perform some particular tweaking on your YunoHost instance?: no
- Using, or trying to install package version/branch: Mastodon 3.5.3~ynh4

### Steps to reproduce
1. Using a browser, go to the Mastodon front page linked above
2. Blank page
3. Same behaviour when logging in to Yunohost and clicking the "Mastodon" tile

### Expected behavior
Using the web front-end should return the login page or the Yunohost SSO login page. Being logged in either way should enable using the Mastodon web front-end.

**Comment by Tagadda** [comment]
This is caused by the `X-Content-Type-Options: nosniff` HTTP header.
![image](https://user-images.githubusercontent.com/36127788/198290616-78ca7490-0cba-4fa1-8aea-b49a25385f34.png)

**Comment by aujawindar** [comment]
Nicely spotted. Is there anything to do about this? I understand that the nosniff HTTP header is a security feature that most browsers use. This is consistent with my experience, as the Mastodon web front-end doesn't load in any browser I've tried.

**Comment by aujawindar** [comment]
Thanks for your help. You are good at finding things that I am not :) But the question remains: Is there a way to solve this specific userland problem?

**Comment by alexAubin** [comment]
If there are reasonable reasons to remove this header, couldnt we just override it with a `more_set_headers` inside Mastodon's nginx conf ?
(Assuming this is the right way to address the issue)

**Comment by Tagadda** [comment]
Ah, yes, but we'll need to do a `more_clear_headers` before `more_set_headers` the to remove the first one.

**Comment by Tagadda** [comment]
Ah, this error was a symptom and not the cause...
![image](https://user-images.githubusercontent.com/36127788/198316496-c1c2eab8-f89d-459d-a0bd-c60130b30d0c.png)
Looks like assets are either not accessible or not compiled ?

Can you share the result of `ls -lah /var/www/mastodon/live/public/packs/css/` ? The file `default-525fdc56.chunk.css` should be present. If not you could try to rebuild the assets with this command : 
`(cd /var/www/mastodon/live && sudo -u mastodon RAILS_ENV=production PATH=/opt/rbenv/versions/mastodon/bin bundle exec rails assets:precompile)`

We also want to check ownership and permissions. The owner should be `mastodon`.

**Comment by aujawindar** [comment]
`ls -lah /var/www/mastodon/live/public/packs/css/` gives
```
drwxr-x--- 2 mastodon www-data 4,0K sep.  20 20:39 .                                                                                    
drwxr-x--- 6 mastodon www-data 4,0K juni  16 00:05 ..                                                                                   
-rw-r----- 1 mastodon www-data 7,0K juni  16 00:05 common-50d0784d.css.gz                                                               
-rw-r----- 1 mastodon www-data 8,3K juni  16 00:05 common-50d0784d.css.map.gz                                                           
-rw-r----- 1 mastodon www-data  76K juni  16 00:05 contrast-770a4110.chunk.css.gz                                                       
-rw-r----- 1 mastodon www-data 172K juni  16 00:05 contrast-770a4110.chunk.css.map.gz                                                   
-rw-r----- 1 mastodon www-data  75K juni  16 00:05 default-525fdc56.chunk.css.gz                                                        
-rw-r----- 1 mastodon www-data 171K juni  16 00:05 default-525fdc56.chunk.css.map.gz                                                    
-rw-r----- 1 mastodon www-data 2,4K juni  16 00:05 mailer-a549806c.chunk.css.gz                                                         
-rw-r----- 1 mastodon www-data  15K juni  16 00:05 mailer-a549806c.chunk.css.map                                                        
-rw-r----- 1 mastodon www-data 3,8K juni  16 00:05 mailer-a549806c.chunk.css.map.gz                                                     
-rw-r----- 1 mastodon www-data  78K juni  16 00:05 mastodon-light-05150082.chunk.css.gz                                                 
-rw-r----- 1 mastodon www-data 878K juni  16 00:05 mastodon-light-05150082.chunk.css.map                                                
-rw-r----- 1 mastodon www-data 178K juni  16 00:05 mastodon-light-05150082.chunk.css.map.gz
```
This means that the files are present, but as gzips?

`(cd /var/www/mastodon/live && sudo -u mastodon RAILS_ENV=production PATH=/opt/rbenv/versions/mastodon/bin bundle exec rails assets:precompile)` gives `bundle: command not found`. Small typo, probably, but I am unable to see what the command should be. Thanks for your help so far.
I have to note that I've done no changes to any Mastodon files. This is a vanilla install.

**Comment by yalh76** [comment]
it should be `(cd /var/www/mastodon/live && sudo -u mastodon RAILS_ENV=production PATH=/opt/rbenv/versions/mastodon/bin bin/bundle exec rails assets:precompile)`

**Comment by aujawindar** [comment]
```
Yarn executable was not detected in the system.                                                                                         
Download Yarn at https://yarnpkg.com/en/docs/install                                                                                    
Yarn executable was not detected in the system.                                                                                         
Download Yarn at https://yarnpkg.com/en/docs/install                                                                                    
sh: 1: node: not found                                                                                                                  
sh: 1: nodejs: not found                                                                                                                
Node.js not installed. Please download and install Node.js https://nodejs.org/en/download/                                              
Exiting!                                                                                                                                
```

I am pretty sure Mastodon was installed through Yarn commands. Also, Yarn and nodejs are already installed.

**Comment by aujawindar** [comment]
To be more precise:
```
# apt show yarn
Package: yarn
Version: 1.22.19-1
```

```
# apt show nodejs
Package: nodejs
Version: 12.22.12~dfsg-1~deb11u1
```

I can't see what's the reason for the rails command to fail with the errors above.

**Comment by nin0-0** [comment]
I checked and yarn path is in PATH, but when I wanted to check the version:
```bash
$ yarn -V
yarn install v1.22.19
[1/6] Validating package.json...
error @mastodon/mastodon@: The engine "node" is incompatible with this module. Expected version ">=12". Got "6.17.1"
error Found incompatible module.
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.

**Comment by aujawindar** [comment]
I should add that this problem was introduced in an update to Mastodon on Yunohost. I can't tell exactly which version introduced the problem, but it was introduced some months ago. Sorry I can't be more specific. The comments to this issue seem to point at problems in the build process.

**Comment by jwgn** [comment]
I can confirm that. Stopped working after some update. I did not notice that immediately, as usage via Mastodon clients (e.g. the official Android client) still works. Just the web-frontend broke.

**Comment by artlog** [comment]
https://github.com/mastodon/mastodon/issues/11600 ?

**Comment by artlog** [comment]
ls -la /usr/local/bin/node 
-rwxr-xr-x 1 root staff 35028569 11 nov.  06:25 /usr/local/bin/node

/usr/local/bin/node --version
v8.17.0

this file does not belong from any package ...

**Comment by jwgn** [comment]
My node version is 16.18.1

```
/usr/local/bin/node --version
v16.18.1
```

**Comment by artlog** [comment]
using PATH=$PATH:/opt/rbenv/versions/mastodon/bin  allows to keep access to nodejs, this was the original problem of script .

<code>
# cd /var/www/mastodon/live && sudo -u mastodon RAILS_ENV=production PATH=$PATH:/opt/rbenv/versions/mastodon/bin 
bin/bundle exec rails assets:precompile
yarn install v1.22.19
[1/6] Validating package.json...
[2/6] Resolving packages...
success Already up-to-date.
Done in 2.00s.
</code>

But still not working.

**Comment by artlog** [comment]
in last resort i did gunzip on directory and restarted mastodon-web

```
cd /var/www/mastodon/live/public/packs/css/
for i in *.*.gz; do gunzip $i; done
```

and for css, js media/fonts media/images ocr ...
Seems better now ...

**Comment by jwgn** [comment]
So is this the recommended way to fix this now? Or can we expect an update where this issue is fixed? I don't want to break it even more.

**Comment by artlog** [comment]
@jwgn i don't have any ground to recommend anything, i just share my way to have it back and running on a production system; i am not working on a fix.

**Comment by jwgn** [comment]
I know. This was more of an indirect question to the maintainers of the package :)

**Comment by jwgn** [comment]
So... what do we do now here?

**Comment by jwgn** [comment]
Issue seems to be fixed with the 4.0.2~ynh1 update. Everything is working again.

**Comment by aujawindar** [comment]
I can confirm. The Yunohost update to Mastodon v4 has fixed the issue for me.
